Recursive dependency detection

mecha opened the pull request:
Solution for #10

mecha committed:
Added recursive dependency detection

Solution for #10

XedinUnknown commented on src/DelegatingContainer.php:
```diff
@@ -38,43 +38,88 @@ public function __construct(ServiceProviderInterface $provider, PsrContainerInte
      */
     public function get($id)
+    {
+        static $stack = [];
```
What is the advantage of using a `static` variable VS property?

mecha replied:
Functionally equivalent.

A static variable makes more sense for me here, since it's an implementation detail of the _function_. A class property would imply that it's some form of state or cache related to the parent class, and I don't consider this temporary stack to be either of those things.

XedinUnknown replied:
Seems like this implementation relies on every actual access to a service being done via an instance of the same class: every time _its_ `get()` method runs, the stack will have a service name pushed onto it. So from my perspective, it is a stack of retrievals from the `DelegatingContainer`, and therefore pertains to the class instance.

XedinUnknown replied:
The bigger problem is that `static` inside a class method will point to the same value _for any instance of that class_. So if somewhere in a service definition a `get()` method of an entirely different `DelegatingContainer` is used, this will push onto the stack.

XedinUnknown replied:
✅ 92c5bfd

XedinUnknown reviewed (changes requested):
Great addition! Just one question.

XedinUnknown committed:
Merge branch '0.2.x' into recursive-detection

# Conflicts:
#	src/DelegatingContainer.php

XedinUnknown committed:
Update lockfile

XedinUnknown committed:
PHPCS will now target PHP 7.4 - the minimal required verson

XedinUnknown committed:
PHPCS will now target PHP 7.4 - the minimal required version

XedinUnknown committed:
Update config

XedinUnknown committed:
Resolution stack now instance-specific

This prevents other unrelated delegating containers from manipulating the stack.

XedinUnknown reviewed (approved):
Very cool!
